File: app/src/main/java/ch/threema/app/activities/SMSVerificationLinkActivity.java
package ch.threema.app.activities;

import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.widget.Toast;

import org.koin.java.KoinJavaComponent;
import org.slf4j.Logger;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import ch.threema.app.R;
import ch.threema.app.di.DependencyContainer;
import ch.threema.app.services.UserService;
import static ch.threema.base.utils.LoggingKt.getThreemaLogger;
import ch.threema.domain.taskmanager.TriggerSource;

import static ch.threema.app.startup.AppStartupUtilKt.finishAndRestartLaterIfNotReady;
import static ch.threema.app.utils.ActiveScreenLoggerKt.logScreenVisibility;

public class SMSVerificationLinkActivity extends AppCompatActivity {
    private static final Logger logger = getThreemaLogger("SMSVerificationLinkActivity");

    @NonNull
    private final DependencyContainer dependencies = KoinJavaComponent.get(DependencyContainer.class);

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        logScreenVisibility(this, logger);
        if (finishAndRestartLaterIfNotReady(this)) {
            return;
        }

        Integer resultText = R.string.verify_failed_summary;

        var userService = dependencies.getUserService();
        if (userService.getMobileLinkingState() == UserService.LinkingState_PENDING) {
            Intent intent = getIntent();
            Uri data = intent.getData();
            if (data != null) {
                final String code = data.getQueryParameter("code");

                if (code != null && !code.isEmpty()) {
                    resultText = null;

                    new AsyncTask<Void, Void, Boolean>() {
                        @Override
                        protected Boolean doInBackground(Void... params) {
                            try {
                                userService.verifyMobileNumber(code, TriggerSource.LOCAL);
                                return true;
                            } catch (Exception e) {
                                logger.error("Failed to verify mobile number", e);
                            }
                            return false;
                        }

                        @Override
                        protected void onPostExecute(Boolean result) {
                            showConfirmation(result ? R.string.verify_success_text : R.string.verify_failed_summary);
                        }
                    }.execute();
                }
            }
        } else if (userService.getMobileLinkingState() == UserService.LinkingState_LINKED) {
            // already linked
            resultText = R.string.verify_success_text;
        } else if (userService.getMobileLinkingState() == UserService.LinkingState_NONE) {
            resultText = R.string.verify_failed_not_linked;
        }

        showConfirmation(resultText);
        finish();
    }

    private void showConfirmation(Integer resultText) {
        if (resultText != null) {
            @StringRes int resId = resultText;

            Toast.makeText(getApplicationContext(), resId, Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/java/ch/threema/app/activities/SMSVerificationLinkActivity.kt
package ch.threema.app.activities

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import ch.threema.android.ToastDuration
import ch.threema.android.showToast
import ch.threema.app.R
import ch.threema.app.services.UserService
import ch.threema.app.startup.finishAndRestartLaterIfNotReady
import ch.threema.app.utils.DispatcherProvider
import ch.threema.app.utils.logScreenVisibility
import ch.threema.base.utils.getThreemaLogger
import ch.threema.domain.taskmanager.TriggerSource
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.koin.android.ext.android.inject

private val logger = getThreemaLogger("SMSVerificationLinkActivity")

class SMSVerificationLinkActivity : AppCompatActivity() {
    init {
        logScreenVisibility(logger)
    }

    private val userService: UserService by inject()
    private val dispatcherProvider: DispatcherProvider by inject()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (finishAndRestartLaterIfNotReady()) {
            return
        }

        lifecycleScope.launch {
            verifyMobileNumber()
            finish()
        }
    }

    private suspend fun verifyMobileNumber() {
        when (userService.getMobileLinkingState()) {
            UserService.LinkingState_PENDING -> {
                val code = intent.data?.getQueryParameter("code")
                if (code.isNullOrEmpty()) {
                    showToast(R.string.verify_failed_summary, ToastDuration.LONG)
                } else {
                    try {
                        withContext(dispatcherProvider.worker) {
                            userService.verifyMobileNumber(code, TriggerSource.LOCAL)
                        }
                        showToast(R.string.verify_success_text, ToastDuration.LONG)
                    } catch (e: Exception) {
                        logger.error("Failed to verify mobile number", e)
                        showToast(R.string.verify_failed_summary, ToastDuration.LONG)
                    }
                }
            }
            UserService.LinkingState_LINKED -> {
                showToast(R.string.verify_success_text, ToastDuration.LONG)
            }
            UserService.LinkingState_NONE -> {
                showToast(R.string.verify_failed_not_linked, ToastDuration.LONG)
            }
        }
    }
}
